docs/mergeop/1_mergeop.md

```diff
@@ -0,0 +1,255 @@
+
+### Proto
+
+The proposed additions to the LLB proto are:
+
+```
+message MergeConflictOpt {
+    bool AllowDeletions = 0;
+    bool AllowFileOverwrites = 1;
+    bool AllowDirectoryOverwrites = 2;
+}
+
+message MergeInput {
+	int64 input = 1;
+}
+
+message MergeOp {
+	repeated MergeInput inputs = 1;
+	MergeConflictOpt mergeConflictOpt = 2;
+}
+```
+
+The structure of MergeOp on the proto level is similar to that of ExecOp, just much simpler since there are less options. MergeInput is comparable to ExecOp’s Mount type; it’s just a reference to an Op.Input, which indexes an input state that should be merged. The MergeOp message type itself is currently just a list of those inputs as no other options are needed for now.
+
+The MergeConflictOpt field will be discussed later in the Conflict Detection doc.
+
+Given their similar structure, MergeOp should be forward compatible in the same way ExecOp is forward compatible today.
+
+## Solver
+
+The solver package will need updates to support the new MergeOp type, namely in solver/llbsolver/ops/. The two methods it will need to implement are:
+1. CacheMap - this will be like a simpler version of ExecOp’s CacheMap method where the cache key is based on the digest of the MergeOp proto and the content checksums of each input to the Merge
+2. Exec - this will just need to use cache.Manager to merge the input refs together (using a new API described in the next sections) and return that ImmutableRef as a Result.
+
+## Cache Modifications
+
+The cache package needs to be modified to support merging ImmutableRefs and to support creating MutableRefs on top of those merged ImmutableRefs. The overall goal here is to make this change as transparent to the rest of the codebase as possible so all existing code that currently uses refs today doesn’t have to be aware of the existence of merged refs in order to use them.
+
+This design thus proposes that we retain the existing ImmutableRef and MutableRef interfaces in the cache package, but update their underlying implementation to handle ref merges.
+* The ImmutableRef.Parent method will need to be replaced with a method that just returns a list of refs for each layer in the chain, but there are only two callers of that method so it’s a very minor update.
+
+The only significant interface change will be extending cache.Accessor with this method:
+* `Merge(ctx context.Context, parents []ImmutableRef, opts ...RefOption) (ImmutableRef, error)`
+
+Merge, as you’d expect, creates an ImmutableRef that represents the merge of the given parents and returns it to the caller. That ImmutableRef can then be used just like any other ImmutableRef can be today.
```

Comment:
If we want to use the "merged" ImmutableRef just like any other ImmutableRef (e.g. using this as a parent of other snapshots), the merged result needs to be recorded to the snapshotter as a snapshot, right? IIUC containerd's overlayfs snaphsotter doesn't support importing a directory as a snapshot, I guess we need to fork our own snapshotter implementation.

Reply:
Surprisingly, we shouldn't need to fork snapshotters; I believe we can implement the whole design on top of any overlay-based snapshotter, including containerd's overlay and even stargz.

For the case you brought up, using a merged ref as the parent of a new snapshot, all of the special handling will take place in Buildkit's cache package. So, if you want to merge together A and B and then create a new snapshot C whose parent is that merged ref, the cache package will create a brand new empty snapshot for C and then merge its mounts on top of the merged mounts of A and B. Essentially, while the end user of LLB thinks of C as a snapshot on top of Merge(A, B), in the underlying snapshotter there are actually 3 separate snapshot chains for A, B and C.

I detailed exactly how this will work in the `Efficient Snapshot Merge` of the `1_mergeop.md` doc (plus the `Efficient Snapshot Merge Details` section in the appendix).

There are some significant corner cases surrounding all this described in the DiffOp and Import/Export docs, but they all have fairly straightforward workarounds thankfully. Let me know if you can think of any I missed though!